### ISSUE TITLE
Provider is unable to assign multiple annotations(tags) to any given object

### ISSUE THREAD

**Issue by kalenarndt** [CiscoDevNet/terraform-provider-aci#734]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

In the UI you are allowed to assign multiple tags (annotations) to a given resource but the provider only allows for a single string to be passed into the resource. This should allow for a list of strings so that way multiple tags can be assigned. 

#### Justification:

Consul-Terraform-Sync will pass the tag values of services in the central service registry and most of the time services have multiple tags associated with them. With the current limitation, we have to pick the first tag in a list and assign it to the resource. If that resource goes down and it has a unique tag it will force an annotation reassignment for all objects created based on the tag in that position. 

### New or Affected Resource(s) + ACI Class(es):

* aci_filter_entry
* aci_contract
* aci_filter
* aci_contract_subject

It's probably all resources since it looks like everything supports the optional annotation flag

### APIC version and APIC Platform

* v4.2(2f) and on-prem/VMware Integration(VMM)

### Potential Terraform Configuration

```hcl
resource "aci_filter" "cts_filter" {
  for_each    = local.service_name
  tenant_dn   = data.aci_tenant.aci_tenant.id
  annotation = one(distinct(flatten(each.value.*.tags))) # ACI provider only supports a single string (non-list)
  name        = "${var.cts_prefix}${each.key}"
  description = "ACI Filter for ${each.key}"
}

resource "aci_filter_entry" "cts_filter_entry" {
  for_each    = local.service_name
  filter_dn   = aci_filter.cts_filter[each.key].id
  name        = "${var.cts_prefix}${each.key}"
  ether_t     = var.ethertype
  prot        = var.protocol
  annotation  = one(distinct(flatten(each.value.*.tags))) # ACI provider only supports a single string (non-list)
  stateful    = "yes"
  d_from_port = tostring(one(distinct(each.value.*.port))) # ACI only supports a single starting port
  d_to_port   = tostring(one(distinct(each.value.*.port))) # ACI only supports a single ending port
}

resource "aci_contract_subject" "cts_contract_subject" {
  for_each                     = local.service_name
  contract_dn                  = aci_contract.cts_contract[each.key].id
  name                         = "${var.cts_prefix}${each.key}_subject"
  relation_vz_rs_subj_filt_att = [aci_filter.cts_filter[each.key].id]
}

resource "aci_contract" "cts_contract" {
  for_each    = local.service_name
  tenant_dn   = data.aci_tenant.aci_tenant.id
  annotation = one(distinct(flatten(each.value.*.tags))) # ACI provider only supports a single string (non-list)
  name        = "${var.cts_prefix}${each.key}_contract"
  description = "Contract for ${each.key}"
  scope       = "application-profile"
}
```

**Comment by maty0609** [comment]
My aci-go-client PR to add annotation to objects has been approved this week. Happy to start working on this.

Some thoughts:
1) Cisco changed annotations from simple string to key:value annotations. This will be breaking change for anyone using annotation strings for ACI version <4.2(4).
2) It will impact all resources.

**Comment by lhercot** [comment]
I think there is a misconception around annotations so let me try to clear it.

The annotation attribute of objects is used to set some values that are moslty used by UI to show special states of objects such as object managed by an external orchestrator like NDO or object that should be invisible in the UI or shows as disabled. We are using this field in the provider mainly to show that it is managed by Terraform (in recent version you see a Terraform label on them). This was always a string that allow a json notation (converted to string).

Depending on the version of your ACI fabric, tags can be represented wither by tagInst objects or by tagAnnotations objects. There is also the tagTag objects which are used for internal matching of objects for things like ESGs or other new features we are looking at.

Those are represented as object children of the object they are assigned to.

We need dedicated resources for them. As I adapted and merged the aci-go-client code last week, I also have the base for the tagAnnotation and tagTag resources. I am not sure we need the tagInst resource as this class is being deprecated in recent version of ACI. By having proper resource for each tag objects we will be able to use the looping syntax to create more than 1 per CTS tag.

**Comment by maty0609** [comment]
Thank you for the clarification. It makes sense. Let me know if you will need any assistance on building new resources.

**Comment by lhercot** [comment]
I have merged the code for the tagTag object (aci_tag) and the tagAnnotation object (aci_annotation) in the main branch and the released v1.2.0 of the provider which include those changes.

You can find the documentation for the aci_annotation resource here: https://registry.terraform.io/providers/CiscoDevNet/aci/latest/docs/resources/annotation

Feel free to open new issues and enhancement requests based on your need.

@maty0609 , thanks for your contribution on the aci-go-client and feel free to propose any other you feel. 
We are working on a developer guide with some scripts we use to generate some of the base code for resources and datasources in the near future to help other contributes.
